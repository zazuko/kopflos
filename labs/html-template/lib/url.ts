import { expand } from '@zazuko/prefixes'
import type { Environment } from '@rdfjs/environment/Environment.js'
import type { DataFactory } from '@rdfjs/types'
import type { NamespaceBuilder } from '@rdfjs/namespace'

export function toNamedNode(env: Environment<DataFactory>, ns: NamespaceBuilder, url: string) {
  if (/^(urn:|https?:\/\/)/.test(url)) {
    return env.namedNode(url)
  }
  const expanded = expand(url)
  if (expanded) {
    return env.namedNode(expanded)
  }

  return ns(url)
}
