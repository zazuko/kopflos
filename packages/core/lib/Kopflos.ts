import type { IncomingHttpHeaders, OutgoingHttpHeaders } from 'node:http'
import type { DatasetCore, NamedNode, Stream } from '@rdfjs/types'
import type { GraphPointer, MultiPointer } from 'clownface'
import type { Options as EndpointOptions, StreamClient } from 'sparql-http-client/StreamClient.js'
import type { ParsingClient } from 'sparql-http-client/ParsingClient.js'
import { CONSTRUCT } from '@tpluscode/sparql-builder'
import { IN } from '@tpluscode/sparql-builder/expressions'
import type { KopflosEnvironment } from './env/index.js'
import { createEnv } from './env/index.js'
import type { ResourceShapeLookup, ResourceShapeMatch } from './resourceShape.js'
import defaultResourceShapeLookup from './resourceShape.js'
import { responseOr } from './responseOr.js'
import type { ResourceLoader, ResourceLoaderLookup } from './resourceLoader.js'
import { fromOwnGraph, findResourceLoader } from './resourceLoader.js'
import type { Handler, HandlerArgs, HandlerLookup } from './handler.js'
import { loadHandler } from './handler.js'
import type { HttpMethod } from './httpMethods.js'

interface KopflosRequest {
  iri: NamedNode
  method: HttpMethod
  headers: IncomingHttpHeaders
}

type ResultBody = Stream | DatasetCore | GraphPointer | Error
export interface ResultEnvelope {
  body?: ResultBody | string
  status?: number
  headers?: OutgoingHttpHeaders
}
export type KopflosResponse = ResultBody | ResultEnvelope

export interface Kopflos {
  get env(): KopflosEnvironment
  get apis(): MultiPointer
  handleRequest(req: KopflosRequest): Promise<ResultEnvelope>
}

interface Clients {
  stream: StreamClient
  parsed: ParsingClient
}

type Endpoint = string | EndpointOptions | Clients

export interface KopflosConfig {
  sparql: Record<string, Endpoint> & { default: Endpoint }
  codeBase?: string
}

export interface Options {
  dataset?: DatasetCore
  resourceShapeLookup?: ResourceShapeLookup
  resourceLoaderLookup?: ResourceLoaderLookup
  handlerLookup?: HandlerLookup
}

type Dataset = ReturnType<KopflosEnvironment['dataset']>

export default class Impl implements Kopflos {
  readonly dataset: Dataset
  readonly env: KopflosEnvironment

  constructor(private readonly config: KopflosConfig, private readonly options: Options = {}) {
    this.env = createEnv(config)

    this.dataset = this.env.dataset([
      ...options.dataset || [],
    ])
  }

  get graph() {
    return this.env.clownface({ dataset: this.dataset })
  }

  get apis(): MultiPointer {
    return this.graph.has(this.env.ns.rdf.type, this.env.ns.kopflos.Api)
  }

  async handleRequest(req: KopflosRequest): Promise<ResultEnvelope> {
    const result = await responseOr(this.findResourceShape(req.iri), (resourceShapeMatch: ResourceShapeMatch) => {
      const resourceShape = this.graph.node(resourceShapeMatch.resourceShape)

      return responseOr(this.findResourceLoader(resourceShape), loader => {
        const coreRepresentation = loader(req.iri, this)

        return responseOr(this.loadHandler(req.method, resourceShapeMatch, coreRepresentation), async handler => {
          const resourceGraph = this.env.clownface({
            dataset: await this.env.dataset().import(coreRepresentation),
          })
          const args: HandlerArgs = {
            resourceShape,
            env: this.env,
            subject: resourceGraph.node(resourceShapeMatch.subject),
            property: undefined,
            object: undefined,
          }
          if ('property' in resourceShapeMatch) {
            args.property = resourceShapeMatch.property
            args.object = resourceGraph.node(resourceShapeMatch.object)
          }

          return handler(args)
        })
      })
    })

    if (this.isEnvelope(result)) {
      return result
    }

    return {
      status: 200,
      body: result,
    }
  }

  async findResourceShape(iri: NamedNode): Promise<ResourceShapeMatch | KopflosResponse> {
    const resourceShapeLookup = this.options.resourceShapeLookup || defaultResourceShapeLookup

    return responseOr(resourceShapeLookup(iri, this), async candidates => {
      if (candidates.length === 0) {
        return { status: 404 }
      }

      if (candidates.length > 1) {
        return new Error('Multiple resource shapes found')
      }

      return candidates[0]
    })
  }

  async findResourceLoader(resourceShape: GraphPointer): Promise<ResourceLoader | KopflosResponse> {
    const resourceLoaderLookup = this.options.resourceLoaderLookup || findResourceLoader

    const loader = await resourceLoaderLookup(resourceShape, this.env)

    return loader || fromOwnGraph
  }

  async loadHandler(method: HttpMethod, resourceShapeMatch: ResourceShapeMatch, coreRepresentation: Stream): Promise<Handler | KopflosResponse> {
    const handlerLookup = this.options.handlerLookup || loadHandler

    const handler = await handlerLookup(resourceShapeMatch, method, this)

    if (handler) {
      return handler
    }

    if (!('property' in resourceShapeMatch) && (method === 'GET' || method === 'HEAD')) {
      return {
        status: 200,
        body: coreRepresentation,
      }
    }

    return { status: 405 }
  }

  private isEnvelope(arg: KopflosResponse): arg is ResultEnvelope {
    return 'body' in arg || 'status' in arg
  }

  static async fromGraphs(kopflos: Impl, ...graphs: Array<NamedNode | string>): Promise<void> {
    const graphsIris = graphs.map(graph => typeof graph === 'string' ? kopflos.env.namedNode(graph) : graph)
    const quads = CONSTRUCT`?s ?p ?o `
      .WHERE`
        GRAPH ?g {
          ?s ?p ?o
        }
        
        FILTER (?g ${IN(...graphsIris)})
      `.execute(kopflos.env.sparql.default.stream)

    for await (const quad of quads) {
      kopflos.dataset.add(quad)
    }
  }
}
