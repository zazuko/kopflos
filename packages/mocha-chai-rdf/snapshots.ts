/* eslint-disable @typescript-eslint/no-explicit-any */
import chai, { expect } from 'chai'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'
import rdf from '@zazuko/env-node'

declare global {
  /* eslint-disable @typescript-eslint/no-namespace */
  namespace Chai {
    interface Assertion {
      canonical: Assertion
    }
  }
}

chai.use(jestSnapshotPlugin())

chai.use((_chai, utils) => {
  const toMatchSnapshot = (chai.Assertion.prototype as any).toMatchSnapshot

  utils.addMethod(chai.Assertion.prototype, 'toMatchSnapshot', function (this: Chai.Assertion, ...args: any[]) {
    const obj = (this as any)._obj

    if (utils.flag(this, 'dataset-canonical')) {
      // Custom behavior for dataset
      expect(rdf.dataset.toCanonical(obj)).toMatchSnapshot(...args)
    } else {
      // Original behavior
      toMatchSnapshot.apply(this, args)
    }
  })

  utils.addProperty(chai.Assertion.prototype, 'canonical', function (this: Chai.Assertion) {
    utils.flag(this, 'dataset-canonical', true)
  })
})
