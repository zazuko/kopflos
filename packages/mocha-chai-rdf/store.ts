/* eslint-disable no-console,camelcase */
import * as url from 'node:url'
import assert from 'node:assert'
import * as Oxigraph from 'oxigraph'
import type { NamespaceBuilder } from '@rdfjs/namespace'
import rdf from '@zazuko/env-node'
import type { DatasetCore, NamedNode, Quad_Graph } from '@rdfjs/types'
import type { AnyPointer } from 'clownface'
import type { Dataset } from '@zazuko/env/lib/Dataset.js'

declare module 'mocha' {
  interface Context {
    dataset: DatasetCore
    graph: AnyPointer
    store: Oxigraph.Store
  }
}

interface DatasetSourceOptions {
  format?: 'trig' | 'nq'
  loadAll?: boolean
  includeDefaultGraph?: boolean
}

interface GraphSourceOptions {
  format?: 'ttl' | 'nt'
}

type Options = (DatasetSourceOptions | GraphSourceOptions) & {
  baseIri?: string | NamespaceBuilder
}

export function createStore(base: string, options: Options = {}) {
  const format = options.format ?? 'ttl'
  let loadAll = true
  let includeDefaultGraph = false
  if (options.format === 'trig' || options.format === 'nq') {
    loadAll = options.loadAll || false
    includeDefaultGraph = options.includeDefaultGraph || false
  }

  return async function (this: Mocha.Context) {
    const store = new Oxigraph.Store()

    const path = url.fileURLToPath(new url.URL(`${base}.${format}`))

    let dataset: Dataset = await rdf.dataset().import(rdf.fromFile(path))

    let graph: Quad_Graph | undefined
    if (this.currentTest && !loadAll) {
      graph = testGraph(this.currentTest)
      dataset = dataset
        .filter((quad) => {
          return quad.graph.equals(graph) || (includeDefaultGraph && quad.graph.equals(rdf.defaultGraph()))
        })
        .map(function toDefaultGraph(quad) {
          return rdf.quad(quad.subject, quad.predicate, quad.object)
        })
    }

    for (const quad of dataset) {
      store.add(quad)
    }

    function assertNotEmpty() {
      if (loadAll) {
        return assert(dataset.size > 0, 'Test data not found')
      }
      if (graph && !graph.equals(rdf.defaultGraph())) {
        return assert(
          dataset.size > 0,
          `Test data not found in GRAPH <${graph.value}>`,
        )
      }

      return assert(
        dataset.match(null, null, null, graph).size > 0,
        `Test data not found in GRAPH <${graph?.value}>`,
      )
    }

    Object.defineProperty(this, 'dataset', {
      get() {
        assertNotEmpty()
        return dataset
      },
      configurable: true,
    })

    Object.defineProperty(this, 'graph', {
      get() {
        assertNotEmpty()
        return rdf.clownface({ dataset })
      },
      configurable: true,
    })
    Object.defineProperty(this, 'store', {
      get() {
        assertNotEmpty()
        return store
      },
      configurable: true,
    })

    // hack to remove the graph after the test just at the right moment
    const cleanup = this.test?.title.includes('before all') ? after : afterEach
    cleanup(() => {
      /* eslint-disable @typescript-eslint/ban-ts-comment */
      // @ts-ignore
      delete this.dataset
      // @ts-ignore
      delete this.graph
      // @ts-ignore
      delete this.store
    })
  }
}

function testGraph(test: Mocha.Test): NamedNode {
  return rdf.namedNode(encodeURI(test.titlePath().slice(1, -1).map(removeSpaces).join('/')))
}

function removeSpaces(arg: string) {
  return arg.replaceAll(' ', '-')
}
