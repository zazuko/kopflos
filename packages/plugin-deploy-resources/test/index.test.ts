import url from 'node:url'
import * as fs from 'node:fs/promises'
import { promisify } from 'node:util'
import { createEmpty } from 'mocha-chai-rdf/store.js'
import rdf from '@zazuko/env-node'
import { expect, use } from 'chai'
import snapshots from 'mocha-chai-rdf/snapshots.js'
import Kopflos from '@kopflos-cms/core'
import configure from '../index.js'
import inMemoryClients from '../../testing-helpers/in-memory-clients.js'

const baseIri = 'http://example.org'
const ex = rdf.namespace(baseIri + '/')

describe('@kopflos-cms/plugin-deploy-resources', () => {
  use(snapshots)

  let env: Kopflos

  beforeEach(createEmpty)

  describe('onStart', () => {
    beforeEach(function () {
      env = new Kopflos({
        baseIri,
        sparql: {
          default: inMemoryClients(this.rdf),
        },
      })
    })

    context('disabled', () => {
      beforeEach(async () => {
        const plugin = configure({
          enabled: false,
        })

        // when
        await plugin.onStart(env)
      })

      it('does nothing', function () {
        expect(this.rdf.dataset.match(null, null, null, ex())).to.have.property('size', 0)
      })
    })

    context('no paths', () => {
      beforeEach(async () => {
        const plugin = configure({
          paths: [],
        })

        // when
        await plugin.onStart(env)
      })

      it('does nothing', function () {
        expect(this.rdf.dataset.match(null, null, null, ex())).to.have.property('size', 0)
      })
    })

    context('path does not exist', () => {
      beforeEach(async () => {
        const plugin = configure({
          paths: ['foobar'],
        })

        // when
        await plugin.onStart(env)
      })

      it('does nothing', function () {
        expect(this.rdf.dataset.match(null, null, null, ex())).to.have.property('size', 0)
      })
    })

    context('enabled', () => {
      beforeEach(async () => {
        const plugin = configure({
          paths: [url.fileURLToPath(new URL('resources', import.meta.url))],
        })

        // when
        await plugin.onStart(env)
      })

      it('deploys trig', async function () {
        const fooGraph = this.rdf.dataset.match(null, null, null, ex('foo'))

        expect(rdf.dataset.toCanonical(fooGraph)).toMatchSnapshot()
      })

      it('applies base', async function () {
        const barGraph = this.rdf.dataset.match(null, null, null, ex('bar'))

        expect(rdf.dataset.toCanonical(barGraph)).toMatchSnapshot()
      })
    })
  })

  context('watch', () => {
    let plugin: ReturnType<typeof configure>

    beforeEach(function () {
      env = new Kopflos({
        baseIri,
        sparql: {
          default: inMemoryClients(this.rdf),
        },
        watch: [],
      })
    })

    context('enabled', () => {
      beforeEach(async () => {
        plugin = configure({
          paths: [url.fileURLToPath(new URL('resources', import.meta.url))],
        })
      })

      it('redeploys when file changes', async function () {
        const fileToModify = url.fileURLToPath(new URL('resources/bar.ttl', import.meta.url))
        const originalContents = await fs.readFile(fileToModify)

        try {
          // given
          await plugin.onStart(env)

          // when
          await fs.appendFile(fileToModify, '<> a ex:Baz .')
          await promisify(setTimeout)(1000)

          // then
          const barGraph = this.rdf.dataset.match(null, null, null, ex('bar'))
          expect(rdf.dataset.toCanonical(barGraph)).toMatchSnapshot()
        } finally {
          await plugin.onStop(env)
          await fs.writeFile(fileToModify, originalContents)
        }
      })

      it('redeploys when file is created', async function () {
        // given
        const fileToCreate = url.fileURLToPath(new URL('resources/baz.ttl', import.meta.url))

        try {
          await plugin.onStart(env)

          // when
          await fs.writeFile(fileToCreate, 'PREFIX ex: <http://example.org/>\n<> a ex:Baz .')
          await promisify(setTimeout)(1000)

          // then
          const bazGraph = this.rdf.dataset.match(null, null, null, ex('baz'))
          expect(rdf.dataset.toCanonical(bazGraph)).toMatchSnapshot()
        } finally {
          await plugin.onStop(env)
          await fs.unlink(fileToCreate)
        }
      })
    })

    context('disabled', () => {
      beforeEach(async () => {
        plugin = configure({
          paths: [url.fileURLToPath(new URL('resources', import.meta.url))],
          watch: false,
        })
      })

      it('does not react to any changes', async function () {
        // given
        const fileToModify = url.fileURLToPath(new URL('resources/bar.ttl', import.meta.url))
        const barContents = await fs.readFile(fileToModify)
        const fileToCreate = url.fileURLToPath(new URL('resources/baz.ttl', import.meta.url))
        const fileToDelete = url.fileURLToPath(new URL('resources/index.trig', import.meta.url))
        const trigContents = await fs.readFile(fileToDelete)

        try {
          await plugin.onStart(env)

          // when
          await Promise.all([
            fs.appendFile(fileToModify, '<> a ex:Baz .'),
            fs.writeFile(fileToCreate, 'PREFIX ex: <http://example.org/>\n<> a ex:Baz .'),
            fs.unlink(fileToDelete),
          ])
          await promisify(setTimeout)(1000)

          // then
          expect(rdf.dataset.toCanonical(this.rdf.dataset)).toMatchSnapshot()
        } finally {
          await fs.unlink(fileToCreate)
          await fs.writeFile(fileToModify, barContents)
          await fs.writeFile(fileToDelete, trigContents)
        }
      })
    })
  })
})
