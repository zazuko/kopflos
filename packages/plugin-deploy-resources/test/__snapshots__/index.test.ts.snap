// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@kopflos-cms/plugin-deploy-resources onStart enabled applies base 1`] = `
"<http://example.org/bar> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.org/Bar> <http://example.org/bar> .
"
`;

exports[`@kopflos-cms/plugin-deploy-resources onStart enabled deploys trig 1`] = `
"<http://example.org/foo> <http://example.org/bar> \\"bar\\" <http://example.org/foo> .
<http://example.org/foo> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.org/Foo> <http://example.org/foo> .
"
`;

exports[`@kopflos-cms/plugin-deploy-resources watch disabled does not react to any changes 1`] = `
"<http://example.org/bar> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.org/Bar> <http://example.org/bar> .
<http://example.org/bar> <urn:talos:action> <urn:talos:overwrite> <urn:talos:resources> .
<http://example.org/bar> <urn:talos:environmentRepresentation> <urn:talos:default> <urn:talos:resources> .
<http://example.org/foo> <http://example.org/bar> \\"bar\\" <http://example.org/foo> .
<http://example.org/foo> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.org/Foo> <http://example.org/foo> .
<http://example.org/foo> <urn:talos:action> <urn:talos:overwrite> <urn:talos:resources> .
<http://example.org/foo> <urn:talos:environmentRepresentation> <urn:talos:default> <urn:talos:resources> .
<urn:talos:resources> <urn:talos:action> <urn:talos:overwrite> <urn:talos:resources> .
"
`;

exports[`@kopflos-cms/plugin-deploy-resources watch enabled redeploys when file changes 1`] = `
"<http://example.org/bar> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.org/Bar> <http://example.org/bar> .
<http://example.org/bar> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.org/Baz> <http://example.org/bar> .
"
`;

exports[`@kopflos-cms/plugin-deploy-resources watch enabled redeploys when file is created 1`] = `
"<http://example.org/baz> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.org/Baz> <http://example.org/baz> .
"`;
